tests/integration/security/reachability/automtls/reachability_test.go

```diff
@@ -0,0 +1,139 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package automtls
+
+import (
+	"testing"
+
+	"istio.io/istio/tests/integration/security/reachability"
+
+	"istio.io/istio/pkg/test/framework"
+	"istio.io/istio/pkg/test/framework/components/echo"
+	"istio.io/istio/pkg/test/framework/components/environment"
+	"istio.io/istio/pkg/test/framework/components/namespace"
+)
+
+// This test verifies reachability under different authN scenario:
+// - app A to app B using mTLS.
+// - app A to app B using mTLS-permissive.
+// - app A to app B without using mTLS.
+// In each test, the steps are:
+// - Configure authn policy.
+// - Wait for config propagation.
+// - Send HTTP/gRPC requests between apps.
+func TestReachabilityAuto(t *testing.T) {
+	framework.NewTest(t).
+		Run(func(ctx framework.TestContext) {
+
+			rctx := reachability.CreateContext(ctx, g, p)
+			systemNM := namespace.ClaimSystemNamespaceOrFail(ctx, ctx)
+
+			testCases := []reachability.TestCase{
+				{
+					ConfigFile:          "global-mtls-on.yaml",
+					Namespace:           systemNM,
+					RequiredEnvironment: environment.Kube,
+					Include: func(src echo.Instance, opts echo.CallOptions) bool {
+						// Exclude calls to the headless TCP port.
+						if opts.Target == rctx.Headless && opts.PortName == "tcp" {
+							return false
+						}
+
+						// Exclude headless->headless
+						return src != rctx.Headless || opts.Target != rctx.Headless
```

Comment:
why not? actually we have fixed the headless issues. So we should be able to test this nicely. There is a flag in pilot to enable this fix [.._HEADLESS_..]. enable it for the tests and you shouldn't need this check.

Reply:
This is cloned from the existing test. I'd prefer to do this in different PR to be clean.